Make 'parse' always return a Money value

smudge opened the pull request:
Took a look at the new `parse`/`parse!` changes. Really awesome!

I was thinking about the way that `String#to_i` always returns a number, even if it's just 0. (Even `nil#to_i` is 0.) As a developer, you never have to worry about getting anything other than an int. (You only have to worry about sanitizing/validating the original string inputs where applicable, but that's on you.)

In the case of `Monetize.parse` and `#to_money`, I realized that the same principle applies. If it returns `nil`, then we've just replaced obligatory exception-handling with obligatory nil-handling, and I'm not sure that's actually an improvement. My thought was that it is better to return `Money.empty` and be consistent with how all of the other non-numeric string inputs behave.

ct-clearhaus commented:
If you're happy about having `nil` as a "non-value value", then I guess you could have code along the lines of
```ruby
def in_my_currency(input_string)
  money = Monetize.parse(input_string) and money.exchange_to('FOO')
end
```
This PR changes the behaviour to have `Money.empty` be the "non-value value" rather than `nil`; and only `nil` is falsey. (And you cannot write `Money.empty.empty?` :grin:)

I'm just thinking about backwards compatibility, not indicating if this change is to the better or worse; I can argue for both, and disagree with myself :smiley:

(Obligatory ref.: http://www.sandimetz.com/blog/2014/12/19/suspicions-of-nil)

antstorm commented:
Really glad that somebody touched this topic. This was called out before — https://github.com/RubyMoney/money-rails/pull/446.

It seems like there are good use cases for both scenarios — returning `nil` vs always returning an instance of `Money`. When looking at this from `money-rails` perspective — it makes sense to have `nil`, indicating that DB field is just empty (as opposed to value being zero). While if you're using `monetize` directly you are most likely to benefit from non-`nil` result.

So it might be good to accommodate both scenarios. Using either a different method or an option passed in…

smudge commented:
I should be more clear about what this changes. As it stands exceptions are only raised in *extremely* rare cases. As I argued in #74, for 99.99% of inputs `#parse`/`#to_money` will return a money value, and developers may not even realize that they need to handle exceptions (or `nil` with the new version of `parse`).

```ruby
Monetize.parse(nil) #=> $0
Monetize.parse("5 November 1605") #=> $51,605
Monetize.parse("This makes no sense") #=> $0
Monetize.parse("3:35pm") #=> $335
Monetize.parse("$50 and loose change") #=> $50
Monetize.parse("Oh look -- a hyphen!") #=> nil ?????
Monetize.parse("123abc") #=> $0
Monetize.parse("123ABC") #=> nil ?????
```

Hence the comparison to `#to_i`, which has a "non-value value" of `0`. Most of the time the "non-value value" of `#to_money` *is* `Money.empty`, and I think that is a useful feature. You don't actually gain anything from the times where you have to handle exceptions (or `nil`), because in practice it's not a useful signal. What action are you going to take when "123 OMG" blows up or returns `nil` that you wouldn't already take when "123 omg" doesn't?

The fact that `Monetize` will, very rarely, throw an exception or return `nil` due to some internal parsing issue seems more like a gotcha than a feature.

smudge commented:
As far as `money-rails` goes, it currently assumes that `Monetize` will mostly return a Money value, even for nonsensical inputs. For its validations, it tracks the string value as originally entered (with the `[field_name]_money_before_type_cast` accessor) and checks against that.

In the case where you disable validations, the field will generally be set to $0 on bad inputs regardless, unless you also set `allow_nil: true`, make the column nullable, and remove the default value in SQL. Even then, the field will probably be set to $0 , because `money-rails` still just relies on `#to_money` except where it detects that the input string is blank. (In which case it skips using `#to_money` altogether.)

So I think you're safe not adding a feature for returning `nil`. If someone really needs it, they can always use `#parse!` and handle it on their own, which you've also made easier now that exceptions are all under one class. 😄

smudge commented:
Actually, when it comes to `money-rails` and the need for `nil`, what would really make sense is an option on `Monetize` that changes all instances where `Money.empty` is returned to instead return `nil`.

That's quite a bit different than just the cases where a parsing exception is thrown. It would be more about making `Monetize.parse(nil)` and `Monetize.parse("")` return `nil`, which could be useful in that Mongoid case.

antstorm commented:
Will all the examples of nonsense input, I wonder if it would make sense to make the parser stricter? Right now it just strips out whatever it doesn't like and treats the result as a single number. So it would make sense for `Monetize.parse('5 November 1605')` to return `nil`, right?

Another solution would be to return an instance of `NilMoney` or `EmptyMoney` (both being the special case of `Money`), following the Null Object Pattern. That way the result still responds to all the methods you might send to it, but allows you to handle that differently (NOTE: `NilMoney.new != Money.empty`).

I think `nil` as the result with an option to specify default makes the most sense so far — `Monetize.parse('$5', default: Money.empty)`. What do you think?

smudge commented:
I had thought about what the Null Object Pattern could provide here, but in practice I wasn't sure how it would work, or whether it would actually be better than `Money.empty`. For instance, how should `NilMoney#to_s` or `NilMoney#format` work? Should it return an empty string? What's the use case for that? Maybe it would make more sense to extend the output of `Money.empty` so that it responds true to `empty?` or `default?` but otherwise behaves like `Money.new(0)`.

Also, what about `#to_money`? Should it have a different default than `Money.parse`? In my experience, if I see that an object returns true on `respond_to?` with a method like `#to_money` or `#to_i`, I expect it to return that type. If I need to sanitize my inputs I will, but the last thing I want is an exception or a `nil` value. (Honestly, I see that as a feature.) Case in point: `nil.to_money` should not be `nil`, or even `NilMoney`. It should be $0. Because I called `to_money`, so I'm asking for a true money value.

Now, whether or not the strictness of the parsing could be improved is a different issue, regardless of what `Money.parse` and `#to_money` return. That could be a nice-to-have, but I think it could also be treated as separate work and I think there is room to improve it more slowly over time.

smudge commented:
In practice, if you're not sanitizing your inputs, you're gonna have a bad time, no matter how good the parser is. 😉

antstorm commented:
Yeah, `to_money` always returning an instance of `Money` is a valid point. On the other hand when calling `Monetize.parse` I want to tell the difference between it being unable to parse the input and detecting actual zero value. This would be a very good use case for an optional, it's a pity that ruby doesn't have them :)

I think it's important to understand that in some use cases `nil` and `0` are different things. And since there are 3 projects involved (`money`, `monetize` and `money-rails`) we need find the best place to make that distinction.

Here's the current behaviour:

```ruby
>> Money.new nil
=> #<Money fractional:0 currency:USD>
```

```ruby
>> Monetize.parse nil
=> #<Money fractional:0 currency:USD>
```

although:

```ruby
>> Money.new nil, ''
Money::Currency::UnknownCurrency: Unknown currency ''
```

smudge commented:
Makes sense!

I wouldn't expect `Money` to handle strings all that well since it's not a parser, but I definitely might expect `Monetize.parse` to tell me more about whether the parsing succeeded. And `#to_money` honoring `respond_to?` and returning an instance of `Money` also makes sense. Will be interesting to see where all of this lands!

Should I close this PR, or is it still useful to get all of the codepaths onto `Money.empty` for now?

antstorm commented:
I think you can make `to_money` always return `Money` as a part of this PR.

Do you think `nil.to_money` still makes sense? Probably does…

smudge commented:
With this change, `#to_money` will always return `Money`, because `Money.empty` == `Money.new(0)`. That part of the PR is fine by me, as-is.

The question is whether this change makes sense for `.parse` in general. I *think* it is useful, because all I'm doing is unifying the codepaths so that it isn't mixing different kinds of "non-value" values. Going forward, you can decide what "non-value" types to return and how to make that configurable, but I'm happy to leave that for future PRs.

The nice thing about using the `rescue` inside of `.parse` to return the "non-value" value is that you can make the parser stricter by just making `parse!` raise more exceptions. 😄

smudge commented:
Hey, I just updated this PR. It adds an `empty_value` option (both globally and in parse's 3rd argument). I also updated the README to include mention of these changes.

Let me know what you think.

smudge commented:
Closing due to inactivity -- might revisit this at some point, but for now this PR is going away in my spring cleaning sweep.
